Add approaches for leap (#653)
exercism/bash: exercises/practice/leap/.approaches/introduction.md

# Introduction

There are various idiomatic approaches to solve Leap.

> You can use a chain of boolean expressions to test the conditions.

Reviewer: ```suggestion
You can use a chain of Boolean expressions to test the conditions.
```

Reviewer 2: Why would boolean be capitalized? I don't think it is a proper noun.

Reviewer: In “Boolean values”, Boolean is a proper adjective referring to George Boole.

Reviewer: Although if boolean is preferred, it should probably be made consistent. `Boolean` also occurs in the PR.

Reviewer 2: After reading https://english.stackexchange.com/questions/4481/should-the-word-boolean-be-capitalized I agree it should probably be capitalized everywhere it's used here.

## General guidance

The key to solving Leap is to know if the year is evenly divisible by `4`, `100` and `400`.

> For determining that, you will use the [modulo operator][modulo-operator].

Reviewer 2: ```suggestion
To determine that, you will use the [modulo operator][modulo-operator].
```

> ## Approach: Arithmetic expression: chain of boolean expressions

Reviewer: ```suggestion
## Approach: Arithmetic expression: chain of Boolean expressions
```

```bash
year=$1
if (( year % 4 == 0 && (year % 100 != 0 || year % 400 == 0) )); then
    echo true
else
    echo false
fi 
```

For more information, check the [Boolean chain approach][approach-boolean-chain].

## Approach: Arithmetic expression Ternary operator of Boolean expressions

```bash
year=$1
if (( year % 100 == 0 ? year % 400 == 0 : year % 4 == 0 )); then
    echo true
else
    echo false
fi 
```

For more information, check the [Ternary operator approach][approach-ternary-operator].

## Approach: External tools

> bash is naturally a "glue" language, making external tools easy to use.

Reviewer: ```suggestion
Bash is naturally a "glue" language, making external tools easy to use.
```

Calling out to a tool that can manipulate dates would be another approach to take.
GNU `date` is an appropriate tool for this problem.

```bash
year=$1
next_day=$(date -d "$year-02-28 + 1 day" '+%d')
[[ $next_day == "29" ]] && echo true || echo false
```

> Add a day to February 28th for the year and see if the new day is the 29th. For more information, see the [external tools approach][approach-external-tools].

Reviewer 2: ```suggestion
Add a day to February 28th for the year and see if the new day is the 29th.
For more information, see the [external tools approach][approach-external-tools].
```

## Which approach to use?

> - The chain of boolean expressions should be the most efficient, as it proceeds from the most likely to least likely conditions.

Reviewer: ```suggestion
- The chain of Boolean expressions should be the most efficient, as it proceeds from the most likely to least likely conditions.
```
